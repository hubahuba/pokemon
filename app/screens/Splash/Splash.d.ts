import {BerryData} from '@/definitions/data/berries.definition.ts';

export interface SplashViewModel {
  data: BerryData[] | undefined;
  isFetching: boolean;
  isFetched: boolean;
  error: Error | null;
}
