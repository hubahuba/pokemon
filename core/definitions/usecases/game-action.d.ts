import {BerryData, PokemonData} from '@/definitions/usecases/pokemon';

export interface GameActionUseCase {
  iChooseYou: (PokemonData) => PokemonData;
  deletePokemon: (pokemonId: string) => void;
  feedPokemon: (pokemonId: string, berry: BerryData) => PokemonData | undefined;
  evolutionPokemon: (pokemonId: string) => PokemonData | undefined;
}
