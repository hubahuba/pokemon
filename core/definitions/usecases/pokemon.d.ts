import {PokemonData} from '@/definitions/data/pokemon.definition';
import {BerryData} from '@/definitions/data/berries.definition';
import {UseInfiniteQueryResult} from '@tanstack/react-query';
import {
  Exact,
  InputMaybe,
  Pokemon_V2_Pokemon_Bool_Exp,
} from '@/definitions/gql/codegen/graphql';
import Pokemon from '@/usecases/pokemon.ts';

export type QuerySearch = Exact<{
  limit: number;
  offset: number;
  search?: InputMaybe<Pokemon_V2_Pokemon_Bool_Exp> | undefined;
}>;

export interface PokemonStat {
  value: number;
  name?: string;
}

export interface PokemonData {
  id: number;
  name: string;
  baseWeight?: number | null;
  currentWeight?: number | null;
  maxWeight?: number | null;
  nextEvolution?: PokemonData[];
  image: any;
  order?: number | null;
  stats?: PokemonStat[];
  evolutions?: PokemonData[];
  lastEat?: string;
  owned?: boolean;
  ownedId?: string;
}

export interface BerryData {
  id: number;
  name: string;
  firmness?: 'very-soft' | 'soft' | 'hard' | 'very-hard' | 'super-hard';
  image: any;
}

export interface BerryFirmness {
  'very-soft': number;
  soft: number;
  hard: number;
  'very-hard': number;
  'super-hard': number;
}

export interface GetBerriesUseCase {
  data: BerryData[] | undefined;
  isFetching: boolean;
  isFetched: boolean;
  error: Error | null;
}

export interface PokemonUseCaseProps {
  getPokemon: (
    offset: number,
    limit: number,
    search?: string | number,
  ) => UseInfiniteQueryResult<PokemonData[] | undefined>;

  getBerries: () => GetBerriesUseCase;
  getMyPokemon: () => PokemonData[];
}

export interface getBerryScoreResult {
  firmness: string;
  score: number;
}

export interface fetchPokemonResponse {
  lastOffset: number;
  pokemons: Pokemon[];
}
