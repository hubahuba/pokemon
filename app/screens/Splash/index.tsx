import * as React from 'react';
import {View, Text} from 'react-native';
import FastImage from 'react-native-fast-image';
import ViewModel from '~/screens/Splash/view-model.ts';

function Splash() {
  ViewModel();
  return (
    <View className="flex flex-1 p-6">
      <View className="grow justify-center items-center">
        <FastImage
          style={{width: 140, height: 140}}
          source={require('~/assets/images/pokeball.png')}
        />
      </View>
      <View className="flex-none justify-center items-center">
        <Text>Pokemon</Text>
      </View>
    </View>
  );
}

export default Splash;
