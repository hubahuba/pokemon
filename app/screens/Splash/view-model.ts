import {useNavigation} from '@react-navigation/native';
import {SplashViewModel} from '~/screens/Splash/Splash';
import {StackNavigation} from '~/routes/routes';
import services from '@/services.ts';

function ViewModel(): SplashViewModel {
  const {useBerries} = services.useCase.pokemon;
  const {data, isFetching, isFetched, error} = useBerries();
  const navigation = useNavigation<StackNavigation>();

  if (isFetched && data?.length) {
    setTimeout(() => {
      navigation.reset({
        index: 0,
        routes: [{name: 'HomeTab'}],
      });
    }, 500);
  }

  return {data, isFetching, isFetched, error};
}

export default ViewModel;
