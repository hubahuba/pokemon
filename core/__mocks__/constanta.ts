import {
  PokeApiQueryQuery,
  PokeGetBerriesQuery,
} from '@/definitions/gql/codegen/graphql.ts';
import {berryDataMapper, pokemonDataMapper} from '@/utils';

export const dummyData: PokeApiQueryQuery = {
  pokemon_v2_pokemon: [
    {
      id: 1,
      name: 'bulbasaur',
      weight: 69,
      pokemon_v2_pokemonsprites: [
        {
          sprites:
            'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png',
        },
      ],
      pokemon_v2_pokemonstats: [
        {
          base_stat: 45,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'HP',
              },
            ],
          },
        },
        {
          base_stat: 49,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Attack',
              },
            ],
          },
        },
        {
          base_stat: 49,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Defense',
              },
            ],
          },
        },
        {
          base_stat: 65,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Special Attack',
              },
            ],
          },
        },
        {
          base_stat: 65,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Special Defense',
              },
            ],
          },
        },
        {
          base_stat: 45,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Speed',
              },
            ],
          },
        },
      ],
      pokemon_v2_pokemonspecy: {
        pokemon_v2_evolutionchain: {
          pokemon_v2_pokemonspecies: [
            {
              id: 1,
              name: 'bulbasaur',
              order: 1,
              pokemon_v2_pokemons: [
                {
                  weight: 69,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 45,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 49,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 49,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 65,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 65,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 45,
                    },
                  ],
                },
              ],
            },
            {
              id: 2,
              name: 'ivysaur',
              order: 2,
              pokemon_v2_pokemons: [
                {
                  weight: 130,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 60,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 62,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 63,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 60,
                    },
                  ],
                },
              ],
            },
            {
              id: 3,
              name: 'venusaur',
              order: 3,
              pokemon_v2_pokemons: [
                {
                  weight: 1000,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/3.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 82,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 83,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                  ],
                },
                {
                  weight: 1555,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/10033.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 123,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 122,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 120,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                  ],
                },
                {
                  weight: 10000,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/10195.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 82,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 83,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
};

export const dummyNoEvo: PokeApiQueryQuery = {
  pokemon_v2_pokemon: [
    {
      id: 3,
      name: 'venusaur',
      weight: 1000,
      pokemon_v2_pokemonsprites: [
        {
          sprites:
            'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/3.png',
        },
      ],
      pokemon_v2_pokemonstats: [
        {
          base_stat: 80,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'HP',
              },
            ],
          },
        },
        {
          base_stat: 82,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Attack',
              },
            ],
          },
        },
        {
          base_stat: 83,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Defense',
              },
            ],
          },
        },
        {
          base_stat: 100,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Special Attack',
              },
            ],
          },
        },
        {
          base_stat: 100,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Special Defense',
              },
            ],
          },
        },
        {
          base_stat: 80,
          pokemon_v2_stat: {
            pokemon_v2_statnames: [
              {
                name: 'Speed',
              },
            ],
          },
        },
      ],
      pokemon_v2_pokemonspecy: {
        pokemon_v2_evolutionchain: {
          pokemon_v2_pokemonspecies: [
            {
              id: 1,
              name: 'bulbasaur',
              order: 1,
              pokemon_v2_pokemons: [
                {
                  weight: 69,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 45,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 49,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 49,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 65,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 65,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 45,
                    },
                  ],
                },
              ],
            },
            {
              id: 2,
              name: 'ivysaur',
              order: 2,
              pokemon_v2_pokemons: [
                {
                  weight: 130,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 60,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 62,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 63,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 60,
                    },
                  ],
                },
              ],
            },
            {
              id: 3,
              name: 'venusaur',
              order: 3,
              pokemon_v2_pokemons: [
                {
                  weight: 1000,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/3.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 82,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 83,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                  ],
                },
                {
                  weight: 1555,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/10033.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 123,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 122,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 120,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                  ],
                },
                {
                  weight: 10000,
                  pokemon_v2_pokemonsprites: [
                    {
                      sprites:
                        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/10195.png',
                    },
                  ],
                  pokemon_v2_pokemonstats: [
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'HP',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Attack',
                          },
                        ],
                      },
                      base_stat: 82,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Defense',
                          },
                        ],
                      },
                      base_stat: 83,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Attack',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Special Defense',
                          },
                        ],
                      },
                      base_stat: 100,
                    },
                    {
                      pokemon_v2_stat: {
                        pokemon_v2_statnames: [
                          {
                            name: 'Speed',
                          },
                        ],
                      },
                      base_stat: 80,
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
};

export const dummyBerries: PokeGetBerriesQuery = {
  pokemon_v2_berry: [
    {
      id: 1,
      name: 'cheri',
      pokemon_v2_berryfirmness: {
        name: 'soft',
      },
      natural_gift_power: 60,
      pokemon_v2_item: {
        pokemon_v2_itemsprites: [
          {
            sprites:
              'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/items/cheri-berry.png',
          },
        ],
      },
    },
    {
      id: 2,
      name: 'chesto',
      pokemon_v2_berryfirmness: {
        name: 'super-hard',
      },
      natural_gift_power: 60,
      pokemon_v2_item: {
        pokemon_v2_itemsprites: [
          {
            sprites:
              'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/items/chesto-berry.png',
          },
        ],
      },
    },
  ],
};

export const dataPokemon = pokemonDataMapper(dummyData);
export const dataLastPokemon = pokemonDataMapper(dummyNoEvo);
export const dataBerry = berryDataMapper(dummyBerries);
