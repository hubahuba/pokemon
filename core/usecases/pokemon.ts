import 'react-native-get-random-values';
import {
  GetBerriesUseCase,
  PokemonUseCaseProps,
  PokemonData,
  QuerySearch,
} from '@/definitions/usecases/pokemon';
import {pokeAPIQueryDocument} from '@/data/pokemons.ts';
import {PokeGetBerriesQueryDocument} from '@/data/berries.ts';
import {pokemonDataMapper, berryDataMapper, filterMapper} from '@/utils';
import {
  useInfiniteQuery,
  UseInfiniteQueryResult,
  useQuery,
} from '@tanstack/react-query';
import request from 'graphql-request';
import Config from 'react-native-config';
import LocalStorage from '@/adapters/local-storage.ts';

const Pokemon: PokemonUseCaseProps = {
  getPokemon: (
    offset,
    limit,
    search,
  ): UseInfiniteQueryResult<PokemonData[] | undefined> => {
    const filter: QuerySearch = filterMapper(offset, limit, search);

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const query = useInfiniteQuery({
      queryKey: ['getPokemons', filter],
      queryFn: async ({pageParam}: any) => {
        const resp = await request(
          Config.API_URL as string,
          pokeAPIQueryDocument,
          pageParam,
        );
        return {
          lastOffset: pageParam.offset,
          pokemons: pokemonDataMapper(resp),
        };
      },
      initialPageParam: filter,
      getNextPageParam: (nextPage, allPages, lastPageParam) => {
        if (lastPageParam && lastPageParam.limit) {
          lastPageParam.offset = nextPage.lastOffset + lastPageParam.limit;
          return lastPageParam;
        }
        return undefined;
      },
    });

    let pokemonData: PokemonData[] = [];
    if (query && query.data) {
      const pokemons = query.data.pages.map(group => group.pokemons);
      pokemonData = pokemons?.reduce((r, e) => (r.push(...e), r), []);
    }

    // @ts-ignore
    return {
      ...query,
      data: pokemonData,
    };
  },

  getBerries: (): GetBerriesUseCase => {
    const localBerries = LocalStorage.getItem('berries');
    if (localBerries) {
      const data = JSON.parse(localBerries);
      return {
        isFetched: true,
        isFetching: false,
        error: null,
        data,
      };
    }
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const {data, isFetching, isFetched, error} = useQuery({
      queryKey: ['getBerries'],
      queryFn: async () => {
        const resp = await request(
          Config.API_URL as string,
          PokeGetBerriesQueryDocument,
        );
        return berryDataMapper(resp);
      },
    });

    if (data) {
      LocalStorage.setItem('berries', JSON.stringify(data));
    }

    return {
      data,
      isFetching,
      isFetched,
      error,
    };
  },

  getMyPokemon: (): PokemonData[] => {
    const localBerries = LocalStorage.getItem('my-pokemons');
    if (localBerries) {
      return JSON.parse(localBerries);
    }
    return [];
  },
};

export default Pokemon;
