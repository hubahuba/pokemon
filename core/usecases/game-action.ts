import {GameActionUseCase} from '@/definitions/usecases/game-action';
import {PokemonData} from '@/definitions/usecases/pokemon';
import LocalStorage from '@/adapters/local-storage.ts';
import {v4 as uuid} from 'uuid';
import {findNextWeight, getBerryScore} from '@/utils';

const GameAction: GameActionUseCase = {
  iChooseYou: data => {
    const selected: PokemonData = {...data};
    let myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      myPokemons = JSON.parse(myPokemons);
    } else {
      myPokemons = [];
    }
    selected.owned = true;
    selected.ownedId = uuid();
    selected.currentWeight = selected.baseWeight;
    (myPokemons as PokemonData[]).push(selected);
    LocalStorage.setItem('my-pokemons', JSON.stringify(myPokemons));
    return selected;
  },

  deletePokemon: pokemonId => {
    const myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      const pokemons: PokemonData[] = JSON.parse(myPokemons);
      const filtered = pokemons.filter(
        pokemon => pokemon.ownedId !== pokemonId,
      );
      LocalStorage.setItem('my-pokemons', JSON.stringify(filtered));
    }
  },

  feedPokemon: (pokemonId, berry) => {
    const myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      const pokemons: PokemonData[] = JSON.parse(myPokemons);
      const currentPokemon = pokemons.find(
        pokemon => pokemon.ownedId === pokemonId,
      );
      if (currentPokemon) {
        const score = getBerryScore(
          berry,
          currentPokemon.lastEat === berry.firmness,
        );
        currentPokemon.lastEat = score.firmness;
        let weight = (currentPokemon.currentWeight as number) + score.score;
        if (weight < 0) {
          weight = 1;
        } else if (weight >= (currentPokemon.maxWeight as number)) {
          weight = currentPokemon.maxWeight as number;
        }
        currentPokemon.currentWeight = weight;
        const filtered = pokemons.filter(
          pokemon => pokemon.ownedId !== pokemonId,
        );
        filtered.push(currentPokemon);
        LocalStorage.setItem('my-pokemons', JSON.stringify(filtered));
        return currentPokemon;
      }
      return currentPokemon;
    }
    return undefined;
  },

  evolutionPokemon: pokemonId => {
    const myPokemons: PokemonData[] | string | undefined =
      LocalStorage.getItem('my-pokemons');
    if (myPokemons) {
      const pokemons: PokemonData[] = JSON.parse(myPokemons);
      let currentPokemon = pokemons.find(
        pokemon => pokemon.ownedId === pokemonId,
      );
      if (currentPokemon) {
        if (currentPokemon.nextEvolution) {
          const nextEvo = [...currentPokemon.nextEvolution];
          const current = nextEvo.shift();
          current.owned = currentPokemon.owned;
          current.ownedId = currentPokemon.ownedId;
          const nextOfNextEvo = findNextWeight(current, nextEvo);
          if (nextOfNextEvo.length > 0 && nextOfNextEvo[0].id !== current.id) {
            current.nextEvolution = nextOfNextEvo;
            current.maxWeight = nextOfNextEvo[0].baseWeight;
          } else {
            current.maxWeight = current.baseWeight;
          }

          const filtered = pokemons.filter(
            pokemon => pokemon.ownedId !== pokemonId,
          );
          filtered.push(current);
          LocalStorage.setItem('my-pokemons', JSON.stringify(filtered));

          return current;
        }

        return currentPokemon;
      }
      return currentPokemon;
    }
    return undefined;
  },
};

export default GameAction;
