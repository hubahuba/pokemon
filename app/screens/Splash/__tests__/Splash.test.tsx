import React from 'react';
import ViewModel from '~/screens/Splash/view-model';
import {
  waitFor,
  renderHook,
  render,
  screen,
} from '@testing-library/react-native';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {BerryData} from '@/definitions/usecases/pokemon';
import nock from 'nock';
import {dummyBerries} from '@/__mocks__/constanta.ts';
import {mockReset} from '../../../../jest.setup';
import Splash from '~/screens/Splash';

describe('Splash view-model', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        gcTime: Infinity,
      },
    },
  });
  const wrapper = ({children}: {children: any}) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  nock('https://beta.pokeapi.co')
    .post('/graphql/v1beta')
    .reply(200, () => ({
      data: dummyBerries,
    }));

  test('viewModel splash check if data is set', async () => {
    const {result} = renderHook(() => ViewModel(), {wrapper});
    await waitFor(() => {
      const {data} = result.current;
      expect((data as BerryData[])[0].name).toBe('cheri');
    });
  });

  test('viewModel splash check if navigated', async () => {
    jest.useFakeTimers();
    renderHook(() => ViewModel(), {wrapper});
    jest.runOnlyPendingTimers();
    await waitFor(() => {
      expect(mockReset.mock.calls.length).toBe(1);
    });
  });

  test('splash screen showed', async () => {
    render(<Splash />, {wrapper});
    await waitFor(() => {
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
