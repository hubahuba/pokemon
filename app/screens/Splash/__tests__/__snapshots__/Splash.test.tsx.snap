// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splash view-model splash screen showed 1`] = `
<View
  style={
    [
      {
        "display": "flex",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "paddingBottom": 24,
        "paddingLeft": 24,
        "paddingRight": 24,
        "paddingTop": 24,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "flexGrow": 1,
        },
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "height": 140,
            "width": 140,
          },
        ]
      }
    >
      <FastImageView
        defaultSource={null}
        resizeMode="cover"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "flexBasis": "auto",
          "flexGrow": 0,
          "flexShrink": 0,
        },
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
      ]
    }
  >
    <Text>
      Pokemon
    </Text>
  </View>
</View>
`;
